module/userExtension/helper.js

```diff
@@ -211,304 +214,357 @@ module.exports = class UserExtensionHelper {
    * @param {Array} userRolesCSVData
    * @param {Object} userDetails -logged in user details.
    * @param {String} userDetails.id -logged in user id.
+   * @param {Object} tenantAndOrgInfo -tenant and organization information passed from req.headers
    * @returns {Array}
    */
 
-  static bulkCreateOrUpdate(userRolesCSVData, userDetails) {
+  static bulkCreateOrUpdate(userRolesCSVData, userDetails, tenantAndOrgInfo) {
     return new Promise(async (resolve, reject) => {
       try {
         let userRolesUploadedData = new Array();
+        let aggregateKafkaEventPayloads = [];
+        // Pre-fetch all required data
+        const allProgramIds = new Set();
+        const allUserIds = new Set();
 
-        const userRolesArray = await userRolesHelper.list(
-          {
-            status: 'active',
-            isDeleted: false,
-          },
-          {
-            code: 1,
-            title: 1,
-            entityTypes: 1,
-            isAPlatformRole: 1,
-          },
+        //iterating through userRolesCSVData to collect all programIds and userIds
+        for (const csvRow of userRolesCSVData) {
+          const userRole = gen.utils.valueParser(csvRow);
+
+          if (userRole.programs && userRole.programs.length > 0) {
+            userRole.programs.forEach((programId) => allProgramIds.add(programId));
+          }
+
+          if (userRole.user) {
+            allUserIds.add(userRole.user);
+          }
+        }
+
+        if(Array.from(allProgramIds).length == 0){
```

Comment:
@borkarsaish65, I had asked you to check whether allUserIds is empty, just below this check. These are the kinds of validations we should prioritize.